### DISCUSSION TITLE
HIL(human-in-the-loop) process support

### DISCUSSION THREAD

**Discussion by rmqrmqrmq** [langflow-ai/langflow#5221] [Q&A]
Hi,
Currently, only one  'chat input' component can be used in per workflow. Is it possible to use more than one? 
For example, the 'chat output' can be connected to the 'chat input' to achieve HIL(human-in-the-loop) process.
Thanks.

**Comment by anovazzi1** [comment]
hey, right now is impossible, but human in the loop is in our road map, and it will be implemented soon

**Reply by rmqrmqrmq** [reply]
Thanks for reply and great to hear that!
Could you share more info about roadmap or timeline for its implementation? 
Thanks.

**Reply by rmedeiros** [reply]
Hi! Do we have any advancements in this topic? Is there any timeline planned for the HIL feature implementation?

**Reply by dgl452** [reply]
Any new update on this? I think is one of the most important missing thing in langflow 😞

**Reply by srlynch1** [reply]
agree, this is needed

**Reply by Lucky2020** [reply]
Agreed, this is a very important feature.

**Reply by juanvecino** [reply]
Hi, does someone know if there has been any updates to this?

In most flows you do not have all information at once and you would like to ask the user for the rest of the information to continue with the flow

**Reply by zhoulijing18** [reply]
现在实现了吗？ how about?

**Reply by tatwong** [reply]
is this ready?

**Comment by rahulwingify** [comment]
Hey @anovazzi1 any updates on this?

**Reply by kvrobert** [reply]
Hey, any updates on this feature? It’d be super important and really useful.

**Comment by tianzhipeng-git** [comment]
a poc implement #8427

**Comment by jeffpasqua** [comment]
Any updates on how this initiative is progressing?

**Comment by antopolskiy** [comment]
I concur this is a much needed feature

**Reply by flefevre** [reply]
Agent workflows will be accepted if we had the possibility to keep human in the loop where it is mandatory

**Comment by KeepALifeUS** [comment]
While waiting for native HIL support, here's a pattern that works today:

## State-Based Human Intervention Points

```python
state = {
    "workflow_id": "wf_123",
    "current_step": "processing",
    "requires_human_input": False,
    "human_prompt": None,
    "human_response": None,
    "agent_outputs": {}
}

# Agent decides it needs human input
if needs_clarification:
    state["requires_human_input"] = True
    state["human_prompt"] = "Please confirm the customer's address"
    state["current_step"] = "waiting_for_human"
    # Workflow pauses here

# After human responds (via API/webhook)
state["human_response"] = user_input
state["requires_human_input"] = False
state["current_step"] = "resume_processing"
```

## Implementation Approach

1. **Polling endpoint**: Workflow checks state periodically
2. **Webhook**: Human response triggers continuation
3. **State persistence**: Store state in DB/Redis for long pauses

## Why This Works

- **No multiple chat inputs** — Single entry point, state manages flow
- **Any step can pause** — Just update state
- **Resumable** — State persists across sessions
- **Debuggable** — Inspect state to see what's waiting

This is the stigmergy pattern applied to HIL: coordination through shared state. More details: https://github.com/KeepALifeUS/autonomous-agents
